File: modules/secure-recorder/ios/KeyManager.swift
import Foundation
import Security

/**
 * Key management for encryption key operations on iOS/iPadOS
 * 
 * Manages encryption keys stored in Keychain. Generates new keys or retrieves existing
 * ones. Keys are accessible after first device unlock, crucial for background processing.
 * 
 * iOS/iPadOS SPECIFICITY:
 * - Uses Keychain (Secure Enclave when available on supported devices)
 * - Key stored with kSecAttrAccessibleAfterFirstUnlock (accessible after first unlock)
 * - Returns Data (raw key bytes) for CryptoKit compatibility
 * - Uses SecRandomCopyBytes for key generation
 */
class KeyManager {
  private let servicePrefix = "expo.modules.securerecorder"
  
  internal func getOrCreateKey(alias: String) throws -> Data {
    // Try to retrieve key from Keychain
    if let existingKey = retrieveKeyFromKeychain(alias: alias) {
      return existingKey
    }
    
    // Generate new 256-bit (32-byte) key
    var newKey = Data(count: 32)
    let result = newKey.withUnsafeMutableBytes { bytes in
      SecRandomCopyBytes(kSecRandomDefault, 32, bytes.baseAddress!)
    }
    
    guard result == errSecSuccess else {
      throw SecureRecorderError.keychainError("Failed to generate encryption key")
    }
    
    // Store in Keychain with kSecAttrAccessibleAfterFirstUnlock
    try storeKeyInKeychain(alias: alias, key: newKey)
    
    return newKey
  }
  
  private func retrieveKeyFromKeychain(alias: String) -> Data? {
    let query: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: servicePrefix,
      kSecAttrAccount as String: alias,
      kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
      kSecReturnData as String: true
    ]
    
    var result: AnyObject?
    let status = SecItemCopyMatching(query as CFDictionary, &result)
    
    guard status == errSecSuccess,
          let data = result as? Data else {
      return nil
    }
    
    return data
  }
  
  private func storeKeyInKeychain(alias: String, key: Data) throws {
    // Delete existing key if present
    let deleteQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: servicePrefix,
      kSecAttrAccount as String: alias
    ]
    SecItemDelete(deleteQuery as CFDictionary)
    
    // Store new key
    let addQuery: [String: Any] = [
      kSecClass as String: kSecClassGenericPassword,
      kSecAttrService as String: servicePrefix,
      kSecAttrAccount as String: alias,
      kSecValueData as String: key,
      kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock
    ]
    
    let status = SecItemAdd(addQuery as CFDictionary, nil)
    
    guard status == errSecSuccess else {
      throw SecureRecorderError.keychainError("Failed to store key: \(status)")
    }
  }
}

File: modules/secure-recorder/android/src/main/java/expo/modules/securerecorder/KeyManager.kt
package expo.modules.securerecorder

import android.content.Context
import android.security.keystore.KeyGenParameterSpec
import android.security.keystore.KeyProperties
import java.security.KeyStore
import javax.crypto.KeyGenerator
import javax.crypto.SecretKey

/**
 * Key management for encryption key operations on Android
 * 
 * Manages encryption keys stored in AndroidKeyStore. Generates new keys or retrieves
 * existing ones. Keys are accessible after first device unlock, crucial for background processing.
 * 
 * ANDROID SPECIFICITY:
 * - Uses AndroidKeyStore (hardware-backed when available on supported devices)
 * - Key stored with setUnlockedDeviceRequired(true) (accessible after first unlock)
 * - Returns SecretKey object for javax.crypto.Cipher compatibility
 * - Uses KeyGenerator for key generation
 */
class KeyManager(private val context: Context) {
  // Internal methods
  internal fun getOrCreateKey(alias: String): SecretKey {
    val keyStore = KeyStore.getInstance("AndroidKeyStore")
    keyStore.load(null)

    // Try to get existing key
    if (keyStore.containsAlias(alias)) {
      val entry = keyStore.getEntry(alias, null) as? KeyStore.SecretKeyEntry
      if (entry != null) {
        return entry.secretKey
      }
    }

    // Generate new key if it doesn't exist
    val keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
    val keyGenParameterSpec = KeyGenParameterSpec.Builder(
      alias,
      KeyProperties.PURPOSE_ENCRYPT or KeyProperties.PURPOSE_DECRYPT
    )
      .setBlockModes(KeyProperties.BLOCK_MODE_GCM)
      .setEncryptionPaddings(KeyProperties.ENCRYPTION_PADDING_NONE)
      .setKeySize(256)
      .setUserAuthenticationRequired(false)
      .setUnlockedDeviceRequired(true)
      .build()

    keyGenerator.init(keyGenParameterSpec)
    return keyGenerator.generateKey()
  }
}
